src/test/scala/mesosphere/mesos/TaskBuilderPortsTestSuite.scala

```diff
@@ -0,0 +1,840 @@
+        AppDefinition(
+          id = "/product/frontend".toPath,
+          cpus = 1.0,
+          mem = 64.0,
+          disk = 1.0,
+          executor = "//cmd",
+          container = Some(Docker(
+            network = Some(DockerInfo.Network.BRIDGE),
+            portMappings = Some(Seq(
+              PortMapping(
+                containerPort = 8080,
+                hostPort = Some(0),
+                servicePort = 9000,
+                protocol = "tcp",
+                name = Some("http"),
+                labels = Map("VIP" -> "127.0.0.1:8080")
+              ),
+              PortMapping(
+                containerPort = 8081,
+                hostPort = None,
+                servicePort = 9001,
+                protocol = "udp",
+                name = Some("admin"),
+                labels = Map("VIP" -> "127.0.0.1:8081")
+              )
+            ))
+          ))
+        )
+
+      val task: Option[(MesosProtos.TaskInfo, Seq[Option[Int]])] = buildIfMatches(offer, appDef)
+      val (taskInfo, taskPorts) = task.get
+      val discoveryInfo = taskInfo.getDiscovery
+
+      "return a defined task" in { task should be('defined) }
+      "set the discovery info name" in { discoveryInfo.getName should be("frontend.product") }
+      "set discovery info has the correct framework visibility" in {
+        discoveryInfo.getVisibility should be (MesosProtos.DiscoveryInfo.Visibility.FRAMEWORK)
+      }
+      "set the correct port names" in {
+        discoveryInfo.getPorts.getPorts(0).getName should be("http")
+        discoveryInfo.getPorts.getPorts(1).getName should be("admin")
+      }
+      "set correct port protocol" in {
+        discoveryInfo.getPorts.getPorts(0).getProtocol should be("tcp")
+        discoveryInfo.getPorts.getPorts(1).getProtocol should be("udp")
+      }
+      "set correct port numbers" in {
+        discoveryInfo.getPorts.getPorts(0).getNumber should be(taskPorts(0).get)
+        discoveryInfo.getPorts.getPorts(1).getNumber should be(8081)
+      }
+      "add VIP labels" in {
+        discoveryInfo.getPorts.getPorts(0).getLabels.getLabels(0).getKey should be("VIP")
+        discoveryInfo.getPorts.getPorts(0).getLabels.getLabels(0).getValue should be("127.0.0.1:8080")
+        discoveryInfo.getPorts.getPorts(1).getLabels.getLabels(0).getKey should be("VIP")
+        discoveryInfo.getPorts.getPorts(1).getLabels.getLabels(0).getValue should be("127.0.0.1:8081")
+      }
+      "add network-scope labels" in {
+        discoveryInfo.getPorts.getPorts(0).getLabels.getLabels(1).getKey should be("network-scope")
+        discoveryInfo.getPorts.getPorts(0).getLabels.getLabels(1).getValue should be("host")
+        discoveryInfo.getPorts.getPorts(1).getLabels.getLabels(1).getKey should be("network-scope")
+        discoveryInfo.getPorts.getPorts(1).getLabels.getLabels(1).getValue should be("container")
+      }
+    }
+
+    "given an offer with duplicated resources and an app definition" should {
```

Comment:
As far as I can see the only difference to other tests is that we add more resources.